### PR TITLE
Merge from development to the master branch

### PR DESCRIPTION
**Pull request by jeroenvanhattem**
Due to complications during the sprint, we are only delivering a part of the vertical slice. We have succesfully implemented the ability to read data from the accelerometer and the gyroscope and calibrate it on the Arduino, instead of using the MPU9250's offset registers.. Unfortunately, we didn't succeed in implementing the AK8963 module, thus we decided to leave this part out of this pull request.

### CODE REVIEW

File: src/mpu9250_interface.cpp [hunk 1]
```diff
@@ -107,17 +109,49 @@ Coordinate3D MPU9250Interface::getMagnetometerValues() {
...
+            magnetometerValues.setY((static_cast<int16_t>(data[3] << 8) | static_cast<int16_t>(data[2])));
+            magnetometerValues.setZ((static_cast<int16_t>(data[5] << 8) | static_cast<int16_t>(data[4])));
         }
     }
     return values;
 }
 
 void MPU9250Interface::printValuesX_Y_Z(Coordinate3D values) {
-    hwlib::cout << hwlib::right << "X: " << values.getX() << " Y: " << values.getY() << " Z: " << values.getZ() << hwlib::endl;
+    hwlib::cout << " X: " << values.getX() << "\r\n"
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp]: Don't use ```\r\n```, change your PuTTY settings instead so you don't need the ```\r```.

File: src/coordinate3d.hpp [hunk 2]
```diff
@@ -89,6 +89,92 @@ struct Coordinate3D {
...
      */
     int16_t getZ();
+
+    /**
+     * @brief coordinate3D += operator
+     *
+     * Returns a coordinate3D after adding another coordinate3D to it.
+     *
+     * @param rhs : const Coordinate3D &
+     *
```
Comment (vasilishomegas) [inline src/coordinate3d.hpp]: According to the Wiki, this newline shouldn't be here, although I won't reject it here.

File: src/memory_map.hpp [hunk 3]
```diff
@@ -21,8 +21,136 @@ enum addresses : uint8_t {
     MPUREG_I2C_SLV0_DO = 0x63,
     MPUREG_I2C_SLV0_REG = 0x26,
     MPUREG_EXT_SENS_DATA = 0x49,
+    READ_FLAG = 0x80,
```
Comment (vasilishomegas) [inline src/memory_map.hpp:24]: This should be an enum class, not an enum.

File: src/mpu9250_interface.cpp [hunk 4]
```diff
@@ -107,17 +109,49 @@ Coordinate3D MPU9250Interface::getMagnetometerValues() {
...
+    saveAccelerationValues();
+    Coordinate3D accelerationCalibrated = accelerationValues - accelerationCalibrateValues;
+    printAccelerationGravity(accelerationCalibrated);
+}
+
+void MPU9250Interface::printAccelerationGravity(Coordinate3D acceleration) {
+    int x = static_cast<int>((acceleration.getX() * 100) * accelRes);
+    int y = static_cast<int>((acceleration.getY() * 100) * accelRes);
+    int z = static_cast<int>((acceleration.getZ() * 100) * accelRes);
+    hwlib::cout << " X: " << x << "/m/s/100\r\n"
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp]: You don't need the ```\r```.

File: src/mpu9250_interface.cpp [hunk 5]
```diff
@@ -107,17 +109,49 @@ Coordinate3D MPU9250Interface::getMagnetometerValues() {
...
+    saveGyroscopeValues();
+    Coordinate3D gyroscopeCalibrated = gyroscopeValues - gyroscopeCalibrateValues;
+    printGyroscopeDegreeSec(gyroscopeCalibrated);
+}
+
+void MPU9250Interface::printGyroscopeDegreeSec(Coordinate3D gyroscope) {
+    int x = static_cast<int>(gyroscope.getX() * gyroRes);
+    int y = static_cast<int>(gyroscope.getY() * gyroRes);
+    int z = static_cast<int>(gyroscope.getZ() * gyroRes);
+    hwlib::cout << " X: " << x << "/degree/s\r\n"
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp]: ```\r```

File: src/main.cpp [hunk 6]
```diff
@@ -13,17 +13,14 @@ int main() {
...
+
         hwlib::cout << "Gyroscope: \r\n";
-        hwlib::cout << sensor.getGyroscopeValues().getX() << "|" << sensor.getGyroscopeValues().getY() << "|"
-                    << sensor.getGyroscopeValues().getZ() << "\r\n";
-        hwlib::cout << "Magnetometer: \r\n";
-        hwlib::cout << sensor.getMagnetometerValues().getX() << "|" << sensor.getMagnetometerValues().getY() << "|"
-                    << sensor.getMagnetometerValues().getZ() << "\r\n";
+        sensor.printGyroscopeDegreeSec();
 
         hwlib::cout << "\r\n";
```
Comment (vasilishomegas) [inline src/main.cpp]: Get rid of the ```\r```.

**Review by vasilishomegas — CHANGES_REQUESTED** [review]

File: src/memory_map.hpp (hunk 3 above)
Comment (Druyv) [reply src/memory_map.hpp:24]: Done!

File: src/mpu9250_interface.cpp [hunk 7]
```diff
@@ -11,8 +11,6 @@
 MPU9250Interface::MPU9250Interface(hwlib::pin_oc &scl, hwlib::pin_oc &sda, const uint8_t MPUAddr)
     : scl(scl), sda(sda), MPUAddr(MPUAddr), i2c(hwlib::i2c_bus_bit_banged_scl_sda(scl, sda)) {
 
-    hwlib::cout << hwlib::right << "Initializing MPU9250..." << hwlib::endl;
-
     uint8_t data[7] = {0x03}; ///< I2C slave 0 register address for AK8963  data
     uint8_t magnReadFlag[1] = {0x0C | 0x80}, readData[1] = {0x87}, continious[1] = {0x06}, bypass[1] = {0x22};
     /// All the I2C data addresses are pointers, because HWLIB only accepts them as such
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp]: I missed this last time, but some comments aren't formatted properly:
```///< Comment```

File: src/mpu9250_interface.cpp [hunk 8]
```diff
@@ -80,20 +82,20 @@ void MPU9250Interface::saveAccelerationValues() {
...
+    accelerationValues.setZ((static_cast<int16_t>(data[4] << 8) | static_cast<int16_t>(data[5])));
 }
 
 void MPU9250Interface::saveGyroscopeValues() {
-    uint8_t data[12] = {0x3B}; ///< address where MPU data starts
+    uint8_t data[6] = {0x43}; ///< address where MPU data starts
 
     i2c.write(MPUAddr, data, 1); // Write 1 byte to MPU
-    i2c.read(MPUAddr, data, 12); // Read 12
+    i2c.read(MPUAddr, data, 6);  // Read 12
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp]: ```///< Comment```

File: src/mpu9250_interface.cpp [hunk 9]
```diff
@@ -107,17 +109,49 @@ Coordinate3D MPU9250Interface::getMagnetometerValues() {
         uint8_t c = data[6]; // end data read by reading ST2
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp:111]: ```///< Comment```

File: src/mpu9250_interface.cpp [hunk 10]
```diff
@@ -15,14 +15,16 @@ MPU9250Interface::MPU9250Interface(hwlib::pin_oc &scl, hwlib::pin_oc &sda, const
     uint8_t magnReadFlag[1] = {0x0C | 0x80}, readData[1] = {0x87}, continious[1] = {0x06}, bypass[1] = {0x22};
     /// All the I2C data addresses are pointers, because HWLIB only accepts them as such
 
-    i2c.write(AK8963_CNTL1, continious, 1);           // start continious mode AK8963
-    i2c.write(MPUREG_I2C_SLV0_ADDR, magnReadFlag, 1); // Set the I2C slave addres of AK8963 and set for read.
-    i2c.write(MPUREG_I2C_SLV0_REG, data, 1);          // I2C slave 0 register address from where to begin data transfe
-    i2c.write(MPUREG_I2C_SLV0_CTRL, readData, 1);     // Read 6 bytes from the magnetometer
+    i2c.write(static_cast<uint8_t>(Addresses::AK8963_CNTL1), continious, 1); // start continious mode AK8963
```
Comment (vasilishomegas) [inline src/mpu9250_interface.cpp:18]: ```///< Comment```

**Review by vasilishomegas — CHANGES_REQUESTED** [review]
Sorry I missed these last time, but make sure inline comments are properly formatted.

**Review by vasilishomegas — APPROVED** [review]

**Pull request merged**